mining: add getTransactions(ByWitnessID) IPC methods

Sjors opened the pull request:
For Stratum v2 custom job declaration to be bandwidth efficient, the pool can request[^0] only the transactions that it doesn't know about.

The spec doesn't specify how this is achieved, but one method is to call the `getrawtransaction` RPC on each transaction id listed in [DeclareMiningJob](https://stratumprotocol.org/specification/06-Job-Declaration-Protocol?query=DeclareMiningJob#644-declareminingjob-client-server) (or a subset if the pool software maintains a cache). Using RPC is inefficient, made worse by the need to make multiple calls. It also doesn't support queuing by witness id (yet, see #34013).

This PR introduces two new IPC methods:

- `getTransactionsById()`: takes a list of `Txid`'s
- `getTransactionsByWitnessID()`: : takes a list of `Wtxid`'s

Both return a list of serialised transactions. An empty element is returned for transactions that were not found.

Unlike the RPC counterpart, the IPC methods do not take advantage of `-txindex`. This could be done in a followup. For `Wtxid` that would involve adding a `-witnesstxindex`.

I thought about having a single (or overloaded) `getTransactions()` that works with both `Txid` and `Wtxid`, but I prefer that clients are intentional about which one they want.

A unit and functional test cover the new functionality.

Sv2 probably only needs `getTransactionsByWitnessID()`, but it's easy enough to just add both.

To rest with Rust use:
- https://github.com/2140-dev/bitcoin-capnp-types/pull/11


[^0]: there's two reasons the pool requests these transactions: to approve the template and to broadcast the block if a solution is found (the miner will also broadcast via their template provider). See also https://github.com/stratum-mining/sv2-spec/issues/170

Sjors committed:
mempool: add lookup by witness hash

Add a simple test for both the Txid and Wtxid variants of CTxMemPool::get().

Sjors committed:
mining: add getTransactionsByWitnessID() IPC method

Sjors committed:
mining: add getTransactionsByTxID() IPC method

Sjors commented:
Needs rebase after #34003, but I'll do that after there's a multiprocess PR.

Sjors committed:
ipc: Serialize null CTransactionRef as empty Data

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

Sjors commented:
I opened https://github.com/bitcoin-core/libmultiprocess/pull/242 and this PR now pulls that in.

Also rebased and renamed `getTransactions` to `getTransactionsByTxID`.

Oh oops, #34568 got merged right before I pushed this. Will rebase again.

Sjors commented:
`p2p_opportunistic_1p1c.py` failure seems spurious: https://github.com/bitcoin/bitcoin/actions/runs/22222027229/job/64279933038?pr=34020

Sjors commented:
Rebased after #34422 and using the latest https://github.com/bitcoin-core/libmultiprocess/pull/242. I added a separate test for `run_transaction_lookup_test` instead of modifying the existing tests in `test/functional/interface_ipc_mining.py`.

Sjors committed:
test: restart node after IPC option override test

Add cleanup so it doesn't need to be the last test.

sedited commented on src/node/interfaces.cpp:
```diff
@@ -1012,6 +1012,20 @@ class MinerImpl : public Mining
+    std::vector<CTransactionRef> getTransactionsByTxID(const std::vector<Txid>& txids) override
+    {
+        if (!m_node.mempool) return {};
+        LOCK(m_node.mempool->cs);
```
Can you lock after the memory allocations (here and in the wtxid implementation)? It also feels weird that we are locking here and using `GetTransaction`. Might `GetTransaction` be the wrong tool for the query anyway? Seems weird that we would hit the txindex through this function, but not the witness one. I realize you mention this in the PR description, but I am having a hard time rationalizing supporting index queries as part of a mining interface. How about calling `mempool.get(...)` directly?

It also seems like a footgun to call `GetTransaction` from a loop that itself locks the mempool. `GetTransaction` can lock `cs_main`, which could invert the usual locking order.

Sjors replied:
> Can you lock after the memory allocations (here and in the wtxid implementation)? It also feels weird that we are locking here and using GetTransaction

This was also brought up here: https://github.com/bitcoin/bitcoin/pull/34020#discussion_r2965707511

I thought I already dropped the outer lock in `...ByTxID(`, but I hadn't. I've now dropped it for both methods.

> I am having a hard time rationalizing supporting index queries as part of a mining interface.

Addressed that in the main thread, below.

> How about calling mempool.get(...) directly?

I'm trying to avoid low level methods in `interface.cpp`. With the outer lock gone, this should be fine now?

sedited replied:
I think I would prefer a mempool member function that locks and returns the required transactions in one go. Isn't there also a race condition without the lock being held while iterating? You now have the potential of returning a transaction and its replacement, which seems confusing?

If there ever is the need for querying the txindex through this, we can simply add that later, no? Adding flexibility that nobody is currently asking for while paying for a potential locking contention penalty seems not ideal. We'll have other interfaces for non-mining clients eventually.

Sjors replied:
> You now have the potential of returning a transaction and its replacement, which seems confusing?

Confusing yes, though it's just one of many ways a block can be invalid. Getting a consistent mempool image is not the goal of this method. Rather, the goal is to be able to construct a block based on (witness) transaction ids.

The mempool is one source of data, we could also have it iterate the extra pool, orphanage, etc.

Sjors replied:
As of 313aaa0eae4446533b3668360ae0a353cde63312 I ended up dropping `-txindex` support and taking the lock outside the loop again.

sedited commented on src/test/mempool_tests.cpp:
```diff
@@ -7,12 +7,16 @@
 #include <test/util/txmempool.h>
 #include <txmempool.h>
 #include <util/time.h>
+#include <node/transaction.h>
```
Nit (clang-format): This in include is not in sorted order.

Sjors replied:
Fixed

Sjors commented:
@sedited wrote (https://github.com/bitcoin/bitcoin/pull/34020#discussion_r3143425195):

> I am having a hard time rationalizing supporting index queries as part of a mining interface.

Afaik it's not necessary for mining clients.

Non-mining clients may find it useful, as a drop-in replacement for the `getrawtransaction` RPC. That gets into the discussion of #34981. An alternative would be to have a second `getTransactionsById()` method on a different interface that _does_ use the index, but that just seems confusing. 

An argument against including it, is that it slows down misses. Though if that's really an issue, it can be addressed with a boolean `use_index` argument.

Sjors commented:
I wrote:

> Afaik it's not necessary for mining clients.

Maybe there is, though it's more nice-to-have than essential.

Searching the index may be slower than searching the mempool, but it's still a lot faster than having to request the transaction from an upstream miner. This should only happen if the miner tip is behind the pool tip, or on a fork, and there are ways to detect that earlier.

_Update_: but then again, the sv2 spec was modified in https://github.com/stratum-mining/sv2-spec/pull/171 to switch from txid to wtxid, so they don't benefit from the index in any case. So I'm inclined to drop it.

sedited reviewed (approved):
ACK 6e4359d0726e39b50cc81cda4769ed25905b4aee

Sjors commented:
I dropped `-txindex` support and switched to more direct `mempool.get()` since the `GetTransaction` wrappers became a bit overkill. I modified `mempool_tests.cpp` to keep the new coverage without those wrappers.

sedited reviewed (approved):
Re-ACK 313aaa0eae4446533b3668360ae0a353cde63312

enirox001 reviewed (commented):
Concept ACK

I took a close look through this and the design makes sense to me.

The PR adds direct mempool transaction lookup to the mining IPC interface with either `txid` or `wtxid` , and the null-`CTransactionRef` serialization change is what makes it possible to preserve positional results across the IPC boundary when a lookup misses. 

The mempool-side changes are straightforward as well, and the tests cover the main behavior I’d want to see here such as txid/wtxid lookup, misses, and transactions no longer being returned once mined

ViniciusCestarii commented:
tACK 313aaa0

Ran tests and tested via a standalone IPC client (separate process, Cap'n Proto over Unix socket) calling both `getTransactionsByTxID` and `getTransactionsByWitnessID` against a mainnet node.

Verified that each method:
- returns the transaction for a known-present ID
- returns a `nullptr` entry for a not present ID

Sjors commented:
Straight-forward rebase after #34860.

sedited reviewed (approved):
ACK 5a21b0a4fe3142e48172f1fe59bedad1b83b7bd9

ViniciusCestarii reviewed (commented):
ACK 5a21b0a4fe3142e48172f1fe59bedad1b83b7bd9

Sjors commented:
Rebased after #33966.

Sjors commented:
One test change had to be moved to an earlier commit. No diff in the final result.

Sjors commented:
Rebased after #34644:
- bumped `.capnp` ordinals because `@7` got taken
- `ipc: Serialize null CTransactionRef as empty Data`: the test now also calls `submitBlock()` with missing block data
- `test: restart node after IPC option override test`: expanded cleanup to rescan `self.miniwallet`

sedited reviewed (approved):
Re-ACK 9784818442f5f66fff7609025e7d530020d7d613

sedited commented:
Ping for another round of review @ViniciusCestarii @enirox001

ViniciusCestarii reviewed (commented):
Re-ACK  9784818442f5f66fff7609025e7d530020d7d613

nit: the commit "ipc: Serialize null CTransactionRef as empty Data" f16b3613cd06db056eba636106047756dcdc1e74 adds submitBlock(b"") empty data test but it seems out of scope here. Maybe put it in its own commit?

ismaelsadeeq reviewed (approved):
Code review ACK 9784818442f5f66fff7609025e7d530020d7d613

This works, but I wonder why we are returning the txs. Isn't it more efficient to just ask the mempool whether these txs by id/wtxid exist? If they do the node should hold on to them till the chain tip changes.

You can reconstruct the block by requesting missing ones and holding on to them. 

When it's time to submit the block, you just pass the missing ones, and the node should already have the rest?

This eliminates the pool having to mirror the nodes' mempool, is bandwidth efficient and is more elegant.

Sjors commented:
> You can reconstruct the block by requesting missing ones and holding on to them.

> When it's time to submit the block, you just pass the missing ones, and the node should already have the rest?

Pass to what? The `submitBlock()` takes a full block, so a client has to construct it first, for which it needs this method.

I'm working on a new IPC method (and interface) that lets clients construct a `BlockTemplate` based on a list `WtxId`s, and where the client can provide missing entries. That would allow the Job Declarator Server role to drop its mempool mirror, and be more efficient as you suggest. But it's a much bigger lift.

ismaelsadeeq commented:
> Pass to what? The submitBlock() takes a full block, so a client has to construct it first, for which it needs this method.

Yes, that does not mean we can not modify it to support new features if they are strictly better.

> I'm working on a new IPC method (and interface) that lets clients construct a BlockTemplate based on a list of WtxIds, and where the client can provide missing entries. That would allow the Job Declarator Server role to drop its mempool mirror, and be more efficient as you suggest. But it's a much bigger lift.

Nice that's great, I prefer for that approach, curious to see it when it's ready, its a shame that we will have this and then have the clients switch to using this better approach later, will it hurt for them to keep using getrawtransaction while we change the approach to do what you suggest instead ?

Sjors commented:
@ismaelsadeeq `getrawtransaction` doesn't support fetching by witness. It also takes one transaction at a time. And hex serialization is still somewhat slow (albeit not as bad as JSON), see https://github.com/bitcoin/bitcoin/pull/32547#issuecomment-4757144295. So this IPC method should be useful for other clients too.

Sjors commented:
@ismaelsadeeq #35671 is a draft PR for the `TxCollection` design discussed above. As you can see it's fair bit more involved than the change here.

achow101 commented:
ACK 9784818442f5f66fff7609025e7d530020d7d613

jasontireshop56-bit reviewed (commented):
JSON

l0rinc commented on src/node/interfaces.cpp:
```diff
@@ -1008,6 +1008,33 @@ class MinerImpl : public Mining
 
+    std::vector<CTransactionRef> getTransactionsByTxID(const std::vector<Txid>& txids) override
+    {
+        if (!m_node.mempool) return {};
```
Fixed in https://github.com/bitcoin/bitcoin/pull/36132
